zfs list: Allow more fields in ZFS_ITER_SIMPLE mode

allanjude opened the pull request:
If the fields to be listed and sorted by is constrained to those populated by dsl_dataset_fast_stat(), then
zfs list is much faster, as it does not need to open each objset and reads its properties.

A previous optimization by Pawel Dawidek (0cee24064a79f9c01fc4521543c37acea538405f) took advantage of this to make listing snapshot names sorted only by name much faster.

However, it was limited to `-o name -s name`, this work extends this optimization to work with:
  - name
  - guid
  - createtxg
  - numclones
  - inconsistent
  - redacted
  - origin

and to work for filesystems as well as snapshots. 
This could be further extended to any other properties supported by dsl_dataset_fast_stat() or similar, that do not require extra locking or reading from disk.

Additional optimizations could be made by making some internal processes, like listing snapshots for deletion
use this "fast" mode, since only the name needs to be looked up, but that is left as future work.

Signed-off-by: Allan Jude <allan@klarasystems.com>

### Motivation and Context
Make `zfs list -H -o name,guid` faster

### Description
Expand the list of properties that trigger 'simple' mode for the zfs list IOCTL
The addition of `createtxg` to the list of optimized properties also means the default sort order no longer requires fetching properties.

### How Has This Been Tested?
Comparing the speed of `zfs list` of 20000 datasets (304% improvement):

`time zfs list -H -o name -s name > /dev/null`
or
`time zfs list -H -o name,guid > /dev/null`
```
x patched_list_name.time
+ unpatched_list_name.time
    N           Min           Max        Median           Avg        Stddev
x  20          1.35          1.37          1.36         1.361  0.0055250625
+  20           5.5          5.52           5.5        5.5055  0.0068633274
Difference at 95.0% confidence
        4.1445 +/- 0.00398763
        304.519% +/- 0.777466%
        (Student's t, pooled s = 0.00623023)
```

as well as `zfs list -t snapshot` for 10 datasets each with 2000 snapshots (20000 snapshots), the improvement was 378%

`time zfs list -H -t snapshot -o name,guid > /dev/null`
```
x patched_list_snapguid.time
+ unpatched_list_snapguid.time
+------------------------------------------------------------+
    N           Min           Max        Median           Avg        Stddev
x  20          1.56          1.58          1.57          1.57  0.0056195149
+  20          7.49          7.54         7.515         7.517   0.015252265
Difference at 95.0% confidence
        5.947 +/- 0.00735649
        378.79% +/- 0.891556%
        (Student's t, pooled s = 0.0114937)
```

To confirm that these changes did not disturb the original enhancement:

`time zfs list -H -t snapshot -o name -s name > /dev/null`
```
x patched_list_snapname.time
+ unpatched_list_snapname.time
+------------------------------------------------------------+
    N           Min           Max        Median           Avg        Stddev
x  20          1.49          1.52           1.5        1.5015  0.0074515982
+  20          1.49          1.51           1.5        1.4975   0.007163504
No difference proven at 95.0% confidence
```


### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

pjd commented on cmd/zfs/zfs_main.c:
```diff
@@ -3593,6 +3586,14 @@ zfs_do_list(int argc, char **argv)
+	/*
+	 * If we are only going to list snapshot names and sort by name,
+	 * then we can use faster version.
+	 */
```
I think this comment needs updating now.

freqlabs commented on cmd/zfs/zfs_iter.c:
```diff
@@ -213,10 +213,57 @@ zfs_free_sort_columns(zfs_sort_column_t *sc)
+	return (1);
+}
+
+int
```
Please return `boolean_t` from both of these checks for readability. `B_TRUE` and `B_FALSE` are unambiguous.

freqlabs commented on lib/libzfs/libzfs_dataset.c:
```diff
@@ -2714,9 +2724,7 @@ zfs_prop_get(zfs_handle_t *zhp, zfs_prop_t prop, char *propbuf, size_t proplen,
-		str = getprop_string(zhp, prop, &source);
-		if (str == NULL)
-			return (-1);
+		str = (const char *)&zhp->zfs_dmustats.dds_origin;
```
This is probably what's causing the test failures:
```suggestion
		str = zhp->zfs_dmustats.dds_origin;
```

allanjude replied:
The test failure it because `char dds_origin[ZFS_MAX_DATASET_NAME_LEN];` is an empty string, so returns blank, while the test expects this to result in `-`. 

I think what we end up needing is:
```suggestion
		str = (const char *)&zhp->zfs_dmustats.dds_origin;
		if (str != NULL && *str == '\0')
			str = zfs_prop_default_string(prop);
```

freqlabs replied:
I think the change I suggested should still be made for consistency (we don't generally take the address of a string and cast it like this).

The case where `str == NULL` is not possible so it would help to remove that check. Otherwise I'm left thinking about "ok but if `str != NULL` is false we then give a NULL pointer to strlcpy?"

allanjude commented:
I've extended this set of changes, addressing the review feedback, but also broadening the optimization to cover listing filesystems, for a significant savings to 'zfs list -o name,guid'

There are a bunch of places where libzfs's zfs_iter_* functions are used internally, that could be changed to take advantage of this. For example the list of snapshots collected during a `zfs destroy -rv`. I've not done that yet, but will look at it later.

behlendorf commented on .github/workflows/checkstyle.yaml:
```diff
@@ -1,7 +1,6 @@
 name: checkstyle
 
 on:
-  push:
```
Unrelated chunk.

behlendorf commented on lib/libzfs/libzfs_iter.c:
```diff
@@ -210,6 +217,9 @@ zfs_iter_bookmarks(zfs_handle_t *zhp, zfs_iter_f func, void *data)
 		return (0);
 
+	if ((flags & ZFS_ITER_SIMPLE) == ZFS_ITER_SIMPLE)
+		zc.zc_simple = B_TRUE;
```
This appears to be unused.

allanjude commented:
I've rebased this and addressed the outstanding review feedback

problame commented:
zrepl would benefit from this feature if `userrefs` was supported as well.
A quick peek suggests that we load the value from the `dsl_dataset_phys::ds_userrefs_obj` into `dsl_dataset::ds_userrefs`.

https://github.com/zfsonlinux/zfs/blob/ba67d82142bc7034734a49a62998cfc96b1d0038/module/zfs/dsl_dataset.c#L641-L648

I don't have time to check whether `ds_userrefs` is ok to read from the `dsl_dataset_fast_stat()` though.

mmaybee commented on cmd/zfs/zfs_iter.c:
```diff
@@ -212,11 +213,58 @@ zfs_free_sort_columns(zfs_sort_column_t *sc)
 {
-	return (sc != NULL && sc->sc_next == NULL &&
-	    sc->sc_prop == ZFS_PROP_NAME);
+	if (sc == NULL) {
+		/* No sort means we don't need any extra properties */
+		return (B_TRUE);
+	}
```
So, the comment is useful here, but the if-statement is not necessary, as we will just fall through the while-loop below and return B_TRUE anyway. Perhaps this comment could be part of a general comment above the function? Something like:
```
/*
 * Return true if all of the properties to be sorted are populated by
 * dsl_dataset_fast_stat(). Note that sc == NULL (no sort) means we
 * don't need any extra properties, so returns true.
 * /

mmaybee commented on lib/libzfs/libzfs_dataset.c:
```diff
@@ -2722,7 +2745,9 @@ zfs_prop_get(zfs_handle_t *zhp, zfs_prop_t prop, char *propbuf, size_t proplen,
 
 	case ZFS_PROP_ORIGIN:
-		str = getprop_string(zhp, prop, &source);
+		str = zhp->zfs_dmustats.dds_origin;
+		if (*str == '\0')
+			str = zfs_prop_default_string(prop);
 		if (str == NULL)
 			return (-1);
```
I assume it is not possible for zhp->zfs_dmustats.dds_origin to be NULL?

allanjude replied:
Fixed the ordering so we check for NULL before dereferencing.

mmaybee commented on lib/libzfs/libzfs_iter.c:
```diff
@@ -152,7 +157,8 @@ zfs_iter_snapshots(zfs_handle_t *zhp, boolean_t simple, zfs_iter_f func,
 		return (0);
 
-	zc.zc_simple = simple;
+	if ((flags & ZFS_ITER_SIMPLE) == ZFS_ITER_SIMPLE)
+		zc.zc_simple = B_TRUE;
```
Nit: this could be simplified to:
```
zc.zc_simple = (flags & ZFS_ITER_SIMPLE) != 0;
```

mmaybee reviewed (approved):
A few nits, but looks good.

behlendorf commented:
@allanjude can you rebase this against master so we can get a clean(er) test run.

behlendorf commented:
Thanks.  So I haven't looked closely in to the [CI failures](http://build.zfsonlinux.org/builders/FreeBSD%20stable%2F13%20amd64%20%28TEST%29/builds/2869) but they pretty clearly were introduced by this commit.

allanjude commented:
> Thanks. So I haven't looked closely in to the [CI failures](http://build.zfsonlinux.org/builders/FreeBSD%20stable%2F13%20amd64%20%28TEST%29/builds/2869) but they pretty clearly were introduced by this commit.

Turns out when I moved the NULL check based on Mark's feedback, I didn't realize that `zfs_prop_default_string()` could actually return null. Fixing.

behlendorf commented:
Thanks for sorting that it, this looks good.  The remaining two CI failures were unrelated.
